### PR TITLE
feat: add workflow to update param names

### LINKED ISSUE #357: Add sync workflow for config param names
Depends on https://github.com/ubiquity/ubiquibot/issues/327

We are in the process of an active development so we refactor the bot's [config](https://github.com/ubiquity/ubiquibot/blob/development/.github/ubiquibot-config.yml) param names occasionally.

The latest refactors (both implemented and planned):
- move from camel case to kebab case (implemented)
- refactor `chainId` to `networkId` (planned)

Updating config param names for all our [partners](https://github.com/ubiquity/devpool-directory/blob/development/projects.json) is tedious. We should automate this process.

What should be done:
- When https://github.com/ubiquity/ubiquibot/issues/327 is implemented create a github workflow with manual dispatch that takes the default yml config (from https://github.com/ubiquity/ubiquibot/issues/327) and creates PRs with updated config param names for [partner projects](https://github.com/ubiquity/devpool-directory/blob/development/projects.json)

So in the end the workflow should be able to create config update PRs which:
1. Add newly added bot's config params
2. Rename config params (example: `chain-id => evm-network-id`)

### PR DESCRIPTION
**Pull request by EtherealGlow**
Resolves #357

### CODE REVIEW

**Comment by EtherealGlow** [comment]
@rndquu how would you go about updating the param keys?

**Comment by rndquu** [comment]
> @rndquu how would you go about updating the param keys?

I would create an "upgrade config object" similar to:
```js
const upgradeConfigObj = {
  "evm-network-id": ["network-id", "chain-id"],
  etc...
}
```
where keys are current param names and key value is an array of previous param names

So the next time we update some config param name we also modify this "upgrade config object" so that the "upgrade config workflow" was aware of the updates

**Comment by EtherealGlow** [comment]
this might work but only making more mappings is left

**Comment by EtherealGlow** [comment]
@rndquu need some help with making mappings for updating the configs

**Comment by EtherealGlow** [comment]
took some help from good ol gpt

**Comment by rndquu** [comment]
> @rndquu need some help with making mappings for updating the configs

Start with this one:
| Old names  | New name |
| ------------- | ------------- |
| network-id, chain-id  | evm-network-id  |

Also notice that you should attach a [QA issue](https://github.com/ubiquity/ubiquibot#how-to-qa-any-additions-to-the-bot) which shows how the current PR works in your forked repo

**Comment by EtherealGlow** [comment]
ill try to set it up also i already did it for evm-network-id

**Comment by EtherealGlow** [comment]
@rndquu it is not working for me

**Comment by EtherealGlow** [comment]
@rndquu can you help me with [this](https://github.com/EtherealGlow/ubiquibot/actions/runs/5705343140/job/15459912352)

**Comment by rndquu** [comment]
> @rndquu can you help me with [this](https://github.com/EtherealGlow/ubiquibot/actions/runs/5705343140/job/15459912352)

Perhaps you need to set the `pull-request` [permission](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs) in the workflow

**Comment by EtherealGlow** [comment]
@rndquu i still cant seem to figure it out
i gave every permission to github_token

**Comment by rndquu** [comment]
> @rndquu i still cant seem to figure it out i gave every permission to github_token

Try to set `permissions: write-all` for testing purposes

Does it work if you create a PR to your own local organization instead of ubiquity org?

**Comment by EtherealGlow** [comment]
it is trying to create a pr to my other repo

**Comment by EtherealGlow** [comment]
also pay.ubq.fi doesnt have a .github/ubiquibot-config.yml

**Comment by EtherealGlow** [comment]
@rndquu can someone else try to qa this? this should work in theory if the GITHUB_TOKEN works.
[ref](https://github.com/EtherealGlow/ubiquibot/actions/runs/5713848445/job/15480032360)

**Comment by rndquu** [comment]
> @rndquu can someone else try to qa this? this should work in theory if the GITHUB_TOKEN works. [ref](https://github.com/EtherealGlow/ubiquibot/actions/runs/5713848445/job/15480032360)

I would try to run the "update config" workflow for your own organization and debug the workflow in the context of your own organization. If the workflow works fine for your own organization then it should probably work for other ones.

**Comment by EtherealGlow** [comment]
@rndquu the issue is done [qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/5/files)

**Comment by EtherealGlow** [comment]
@0xcodercrane can you review this workflow?
here is the [qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/5/files)

File: .github/workflows/update-config.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
+name: Pull Request Action
+permissions: write-all
+on:
+  push:
```
Comment (rndquu) [inline .github/workflows/update-config.yml:4]: Pls remove the `push` handler, we don't want to spam with PRs

File: .github/workflows/update-config.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          git config --global user.name "github-actions[bot]"
+
+      - name: Install jq and yq
+        run: |
+          sudo apt-get -y install jq
+          sudo wget https://github.com/mikefarah/yq/releases/latest/download/yq_linux_amd64 -O /usr/bin/yq
+          sudo chmod +x /usr/bin/yq
+      - name: Update Config Params and Create Pull Requests
+        run: |
+          #urls=$(curl -sSL https://raw.githubusercontent.com/ubiquity/devpool-directory/development/projects.json | jq -r '.urls[]')
```
Comment (rndquu) [inline .github/workflows/update-config.yml:29]: Shouldn't we uncomment it?

File: .github/workflows/update-config.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+              old_params="${param_mapping[$new_param]}"
+              for old_param in $old_params
+              do
+                yq ".$new_param = .$old_param | del(.$old_param)" .github/ubiquibot-config.yml > .github/ubiquibot-config
+                mv .github/ubiquibot-config .github/ubiquibot-config.yml 
+              done
+            done
+            
+            git add .
+            git commit -m "fix: use latest ubiquibot config setup"
```
Comment (rndquu) [inline .github/workflows/update-config.yml:66]:
```suggestion
            git commit -m "build: use latest ubiquibot config setup"
```

File: .github/workflows/update-config.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            git remote set-url origin https://${{ secrets.PAT_TOKEN }}@github.com/$ownerName/$repoName.git
+            git push -f origin update
+            curl -L \
+            -X POST \
+            -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer ${{ secrets.PAT_TOKEN }}"\
+            -H "X-GitHub-Api-Version: 2022-11-28" \
+             https://api.github.com/repos/$ownerName/$repoName/pulls \
+            -d '{
+             "title": "fix: use latest ubiquibot config setup",
```
Comment (rndquu) [inline .github/workflows/update-config.yml:76]:
```suggestion
             "title": "build: use latest ubiquibot config setup",
```

**Review by rndquu — CHANGES_REQUESTED** [review]

File: .github/workflows/update-config.yml (hunk 1 above)
Comment (EtherealGlow) [reply .github/workflows/update-config.yml:4]: should we do this on workflow_dispatch?

Comment (EtherealGlow) [reply .github/workflows/update-config.yml:4]: or 
```
on:
  push:
    paths:
    - 'ubiquibot-config.-default.json'
```

**Comment by EtherealGlow** [comment]
[qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/3/files)

**Comment by rndquu** [comment]
So we've got the [default json config](https://github.com/ubiquity/ubiquibot/blob/development/ubiquibot-config-default.json) which is 100% relevant (in terms of correct param names) at all times.

Partners, who use the [ubiquibot](https://github.com/ubiquity/ubiquibot), can modify bot parameters setting a special yml config in their own repos. Take this [config](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/ubiquibot-config.yml) as an example. If we compare the [default config](https://github.com/ubiquity/ubiquibot/blob/development/ubiquibot-config-default.json)  and [partner's config](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/ubiquibot-config.yml), we may see, that the `evm-network-id` param is missing in the [partner's config](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/ubiquibot-config.yml). So the workflow (implemented in the current issue) should create a PR to the [partner's config](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/ubiquibot-config.yml) which adds the `evm-network-id` param (along with other missing parameters).

Does this [QA](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/3) show the behavior described above?

**Comment by me505** [comment]
on it

**Comment by EtherealGlow** [comment]
[qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/5/files)
i didnt know where to get the default value from so i took from this config

**Comment by EtherealGlow** [comment]
@rndquu can you review this?

**Review by rndquu — CHANGES_REQUESTED** [review]
How it should work:
1. Take the [default bot's config](https://github.com/ubiquity/ubiquibot/blob/development/ubiquibot-config-default.json)
2. Fetch configs from all [partners](https://github.com/ubiquity/devpool-directory/blob/development/projects.json) (partner's YML config [example](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/ubiquibot-config.yml))
3. Compare the [default json bot's config](https://github.com/ubiquity/ubiquibot/blob/development/ubiquibot-config-default.json) with a [partners' yml configs](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/ubiquibot-config.yml) and create a PR to YML config for the newly added or modified parameters

As far as I understand the workflow implemented in this PR creates PRs to `json` config instead of `yml`, pls correct me if I'm wrong

**Comment by EtherealGlow** [comment]
[qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/8/files)

**Comment by EtherealGlow** [comment]
[qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/9/files)

File: .github/workflows/update-config.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v3
+
+      - name: Set up Git
+        run: |
+          git config --global user.email "github-actions[bot]@users.noreply.github.com"
+          git config --global user.name "github-actions[bot]"
```
Comment (pavlovcik) [inline .github/workflows/update-config.yml:19]: Could be cool to authorize as the UbiquiBot instead. Here's a snippet from some old CI:

https://github.com/ubiquity/ubiquibot/blob/c8d310ec896719dc43f28cee2697bf4648ac0e5b/.github/workflows/push.yml#L29-L45

Comment (pavlovcik) [reply .github/workflows/update-config.yml:19]: Just checked and we still have the secret! This should work still. Would you mind adding it in here?

image

**Review by pavlovcik — CHANGES_REQUESTED** [review]

**Comment by pavlovcik** [comment]
> [qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/9/files)

QA looks great thank you!

**Comment by EtherealGlow** [comment]
@pavlovcik i skipped learning ssh so i dont know if this commit will be ok

**Comment by pavlovcik** [comment]
> @pavlovcik i skipped learning ssh so i dont know if this commit will be ok

Post another QA and we'll find out together!

**Comment by EtherealGlow** [comment]
[qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/10/files)

**Comment by EtherealGlow** [comment]
cant see much of a difference

File: .github/workflows/update-config.yml [hunk 6]
```diff
@@ -22,7 +27,7 @@ jobs:
       - name: Authenticate SSH 
         uses: webfactory/ssh-agent@v0.4.0 
         with: 
-          ssh-private-key: ${{ secrets.UBIQUITY_BOUNTY_BOT_DEV_PRIVATE_KEY_OPENSSH }} 
+          ssh-private-key: ${{ secrets.ssh }}
```
Comment (pavlovcik) [inline .github/workflows/update-config.yml]: That doesn't look right

**Review by pavlovcik — CHANGES_REQUESTED** [review]

**Comment by pavlovcik** [comment]
It doesn't look like you have Actions enabled on your fork so I'm not sure how you can run CI.

File: .github/workflows/update-config.yml (hunk 6 above)
Comment (EtherealGlow) [reply .github/workflows/update-config.yml]: ssh was the secret for my own fork

**Comment by 0xcodercrane** [comment]
I am also not sure whether this issue is still valuable to do

**Comment by EtherealGlow** [comment]
but the issue is almost completed tho

**Comment by rndquu** [comment]
@EtherealGlow I've slightly modified the [workflow](https://github.com/rndquu/bounty-bot/actions/runs/5835145152/workflow) for testing purposes and get this [error](https://github.com/rndquu/bounty-bot/actions/runs/5835145152/job/15826303480).

Any ideas of what I'm doing wrong?

**Comment by rndquu** [comment]
> I am also not sure whether this issue is still valuable to do

Running a workflow is better than manually updating the params every week

**Comment by EtherealGlow** [comment]
try web factory ssh agent and add the ssh key to your repo

**Comment by EtherealGlow** [comment]
I don't get as to why we can't use https

**Comment by EtherealGlow** [comment]
> @EtherealGlow I've slightly modified the [workflow](https://github.com/rndquu/bounty-bot/actions/runs/5835145152/workflow) for testing purposes and get this [error](https://github.com/rndquu/bounty-bot/actions/runs/5835145152/job/15826303480).
> 
> Any ideas of what I'm doing wrong?

maybe you haven't used web factory ssh agent?
idk

**Comment by EtherealGlow** [comment]
> @EtherealGlow I've slightly modified the [workflow](https://github.com/rndquu/bounty-bot/actions/runs/5835145152/workflow) for testing purposes and get this [error](https://github.com/rndquu/bounty-bot/actions/runs/5835145152/job/15826303480).
> 
> Any ideas of what I'm doing wrong?

maybe you haven't used web factory ssh agent?
idk

**Comment by rndquu** [comment]
> > @EtherealGlow I've slightly modified the [workflow](https://github.com/rndquu/bounty-bot/actions/runs/5835145152/workflow) for testing purposes and get this [error](https://github.com/rndquu/bounty-bot/actions/runs/5835145152/job/15826303480).
> > Any ideas of what I'm doing wrong?
> 
> maybe you haven't used web factory ssh agent to set up ssh? idk

I can't find any occurrences of the "Pull Request Action" workflow in the [workflow run history](https://github.com/EtherealGlow/ubiquibot/actions?query=branch%3Adevelopment-3) of your repo

Could you show the workflow run that created a PR for the QA you mentioned [here](https://github.com/ubiquity/ubiquibot/pull/562#issuecomment-1673608609)?

**Comment by EtherealGlow** [comment]
https://github.com/EtherealGlow/ubiquibot/actions/runs/5824095655/job/15792393412

File: .github/workflows/update-config.yml [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v3
+      
+      - name: Set up git
+        run: |
+          git config --global user.email "UbiquiBot[dev]@github.com"
```
Comment (pavlovcik) [inline .github/workflows/update-config.yml:18]:
```suggestion
          git config --global user.email "UbiquiBot[bot]@github.com" 
```
I think that's the correct username for the associated secret

**Comment by pavlovcik** [comment]
Can you post proof that it works to commit as UbiquiBot?

**Comment by EtherealGlow** [comment]
@rndquu how would you go on about change commiter info

**Comment by pavlovcik** [comment]
Refer to [this](https://github.com/ubiquity/ubiquibot/pull/562#discussion_r1289064260) and figure the rest out with ChatGPT etc

**Comment by EtherealGlow** [comment]
@pavlovcik removing ssh made it work somehow

**Comment by EtherealGlow** [comment]
[qa](https://github.com/EtherealGlow/ubiquibot-pr-qa/pull/16)

**Comment by EtherealGlow** [comment]
huh?

**Pull request closed without merging**